examples/gno.land/p/demo/subscription/subscription.gno

```diff
@@ -0,0 +1,123 @@
+// Package subscription provides a library for managing different types of
+// subscriptions in Gno applications. It supports both recurring and
+// lifetime subscriptions, enabling users to access services based on their
+// subscription status. This package uses a tree-based data structure to
+// efficiently track and manage subscription statuses.
+
+// Example Usage:
+//
+//	import "gno.land/p/demo/subscription"
+//
+//	Create a recurring subscription for 30 days costing 100 units.
+//	recurringSub := subscription.NewRecurringSubscription(time.Hour*24*30, 100)
+//
+//	Create a lifetime subscription for a one-time payment of 500 units.
+//	lifetimeSub := subscription.NewLifetimeSubscription(500)
+//
+//	func HandleRequest(caller std.Address) {
+//		Check access for a recurring subscription.
+//		recurringSub.CheckAccess(caller)
+//
+//		Perform payment for a lifetime subscription.
+//		lifetimeSub.ProcessPayment(caller, 500)
+//	}
+
+package subscription // import "gno.land/p/demo/subscription"
+
+import (
+	"gno.land/p/demo/ufmt"
+	"gno.land/p/demo/avl"
+	"time"
+	"std"
+)
+
+// Subscription interface defines standard methods that all subscription types must implement.
+type Subscription interface {
+	CheckAccess(std.Address)
+	ProcessPayment(std.Address, int64)
+}
+
+// RecurringSubscription represents a subscription that requires periodic payments.
+// It includes the duration of the subscription and the amount required per period.
+type RecurringSubscription struct {
+	duration time.Duration 
+	amount   int64
+	subs     *avl.Tree      // std.Address -> time.Time
+}
+
+// LifetimeSubscription represents a subscription that requires only a one-time payment.
+// It grants permanent access to a service or product.
+type LifetimeSubscription struct {
```

Comment:
What do you think about adding a method to update the subscription amount? Both for the lifetime subscription and recurring one. It can be a method on the `Subscription` interface, and in the implementation you can guard against it with Ownable or something similar